scipy/optimize/_qap.py

```diff
@@ -0,0 +1,786 @@
+    if shuffle_input:
+        seed_dist_c = rng.permutation(seed_dist_c)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    seed_cost_c = np.setdiff1d(range(n), partial_match[:, 0])
+    perm_A = np.concatenate([partial_match[:, 0],
+                             seed_cost_c], axis=None).astype(int)
+    perm_B = np.concatenate([partial_match[:, 1],
+                             seed_dist_c], axis=None).astype(int)
+
+    A = A[np.ix_(perm_A, perm_A)]
+    B = B[np.ix_(perm_B, perm_B)]
+
+    # definitions according to Seeded Graph Matching [2].
+    A11 = A[:n_seeds, :n_seeds]
+    A12 = A[:n_seeds, n_seeds:]
+    A21 = A[n_seeds:, :n_seeds]
+    A22 = A[n_seeds:, n_seeds:]
+    B11 = B[:n_seeds, :n_seeds]
+    B12 = B[:n_seeds, n_seeds:]
+    B21 = B[n_seeds:, :n_seeds]
+    B22 = B[n_seeds:, n_seeds:]
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    if init_weight is not None:
+        # generate a nxn matrix where each entry is a random number [0, 1]
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+    grad_P = np.inf  # gradient of P
+    n_iter = 0  # number of FW iterations
+
+    # OPTIMIZATION WHILE LOOP BEGINS
+    while grad_P > tol and n_iter < maxiter:
+        # computing the gradient of f(P) = -tr(APB^tP^t)
+        delta_f = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+        # run hungarian algorithm on gradient(f(P))
+        rows, cols = linear_sum_assignment(obj_func_scalar * delta_f)
+        Q = np.zeros((n_unseed, n_unseed))
+        Q[rows, cols] = 1  # initialize search direction matrix Q
+
+        def f(x):  # computing the original optimization function
+            xP1xQ = x * P + (1 - x) * Q
+            # Sums below are np.trace(A11.T @ B11)
+            # + np.trace(xP1xQ.T @ A21 @ B21.T)
+            # + np.trace(xP1xQ.T @ A12.T @ B12)
+            # + np.trace(A22.T @ xP1xQ @ B22 @ xP1xQ.T)
+            # This is more efficient, but can we do even better?
+            return obj_func_scalar * (
+                (A11 * B11).sum()
+                + (xP1xQ.T @ A21 * B21).sum()
+                + (xP1xQ.T @ A12.T * B12.T).sum()
+                + ((xP1xQ.T @ A22) * (B22 @ xP1xQ.T)).sum()
+            )
+
+        # computing the step size
+        alpha = minimize_scalar(f, bounds=(0, 1), method="bounded").x
+        P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+        grad_P = np.linalg.norm(P - P_i1)
```

Comment:
Read the FAQ paper, `grad_P` is not a gradient but a change (or delta)

We can just do:
```
if np.linalg.norm(P - P_i1) < tol:
    break
````
and get rid of the variable.

The while loop can then be replaced with:
`for n_iter in range(maxiter):`

Reply:
yeah bad variable naming on my part. I'm for this change.

Reply:
So now we go through the trouble to calculate `P_i1` but only use it to decide to terminate. We might as well update `P` with it before braking, right?